File: temp_img_proc.py
def load_pgm(filename):
    """Load an ASCII PGM (P2) into a 2D list."""
    with open(filename) as f:
        lines = [l for l in f if l.strip() and not l.startswith('#')]
    if lines[0].strip() != 'P2':
        raise ValueError("Only ASCII PGM (P2) supported")
    
    width, height = map(int, lines[1].split())
    pixels = list(map(int, ' '.join(lines[3:]).split()))
    return [pixels[i*width:(i+1)*width] for i in range(height)]
def lr_range(x, step):
    '''Creates a list of [x-step, ..., x, ..., x+step]'''
    res = []

    for i in range(-step, step+1):
        res.append(x + i)
    return res

def avg_pixel(img, y, x, sz):
    half = sz // 2

    sum_window = []
    # xs = list(range(-half, half+1))

    xlo = max(x-half, 0)
    xhi = min(x+half+1, len(img[y]))

    for irow in lr_range(y,half):
        if irow < 0 or irow >= len(img):
            row = []
        else:
            row = img[irow][xlo:xhi]

        sum_window = sum_window + row

    if len(sum_window) == 0:
        return img[y][x]

    return sum(sum_window)/len(sum_window)


def gaussian_multiplier(sz):
    '''returns eg. [1,2,4,2,1]'''
    half = sz // 2
    weights = []
    value = 1
    
    for i in range(sz):
        weights.append(value)
        if i < half:
            value *= 2
        else:
            value //= 2
            
    return weights

def gaussianpixel(img, y, x, sz):
    #Gaussian
    half = sz // 2
    weights = gaussian_multiplier(sz)
    # weights will be the same for x & y
    wd = len(img[0])
    ht = len(img)
    total = 0
    weight_sum = 0

    for dy in range(-half, half + 1):
        imgy = y + dy #range of y 
        if 0 <= imgy < ht:
            wy = weights[dy + half]  # vertical weight
            for dx in range(-half, half + 1):
                imgx = x + dx #range of x
                if 0 <= imgx < wd:
                    wx = weights[dx + half]  # horizontal weight
                    w = wy * wx
                    total += img[imgy][imgx] * w
                    weight_sum += w

    return total / weight_sum if weight_sum != 0 else img[y][x]

def pixeleval(img, template, y, x):

    sz = len(template)
    half = sz // 2

    ht = len(img)
    wd = len(img[0])

    res = 0
    temp_sum = 0

    for yT in range(sz):
        for xT in range(sz):

            ny = y + yT - half
            nx = x + xT - half

            if 0 <= ny < ht and 0 <= nx < wd:
                w = template[yT][xT]
                res += img[ny][nx] * w
                temp_sum += w

    return res / temp_sum if temp_sum != 0 else 0
def img_proc(img,sz, operation="avg", template=None):
    method = 0
    res = []
    ht = len(img)
    wd = len(img[0])
    for y in range(ht):
        res.append([])
        for x in range(wd):
            if template == None:
                if operation == "gaussian":
                    method = gaussianpixel(img,x,y, sz)
                elif operation == 'avg':
                    method = avg_pixel(img,y,x,sz)
            else:
                method = pixeleval(img, template,y,x,)
            res[y].append(method)
    return res

def main():

    img = load_pgm("feep.ascii.pgm")

    while True:

        print("\n=== IMAGE PROCESSING MENU ===")
        print("1. Average Blur")
        print("2. Gaussian Blur")
        print("3. Custom Template")
        print("4. Quit")

        choice = input("Choose an option: ")

        # Average blur
        if choice == "1":

            sz = int(input("Kernel size: "))

            result = img_proc(img, sz, operation="avg")

            print("\nProcessed Image:")
            print(result)

        # Gaussian blur
        elif choice == "2":

            sz = int(input("Kernel size: "))

            result = img_proc(img, sz, operation="gaussian")

            print("\nProcessed Image:")
            print(result)

        # Custom template / convolution matrix
        elif choice == "3":

            sz = int(input("Template size: "))

            print(f"\nEnter {sz * sz} numbers separated by spaces:")

            vals = list(map(int, input().split()))

            # Build 2D template matrix
            template = []

            for i in range(sz):

                row = vals[i * sz:(i + 1) * sz]

                template.append(row)

            print("\nTemplate:")
            for row in template:
                print(row)

            result = img_proc(img, sz, template=template)

            print("\nProcessed Image:")
            print(result)

        # Quit
        elif choice == "4":

            print("Goodbye")
            break

        else:
            print("Invalid option")


if __name__ == "__main__":
    main()


# def img_proc(img, template):
#     ht = len(img)
#     wd = len(img[0])
    
#     res = []
#     for y in range(ht):
#         res.append([])
#         for x in range(wd):
#             res[y].append(pixeleval(img, template, y, x))
#     return res

File: temp_img_proc.js
const fs = require('fs');

function range(start = 0, end) {
    const arr = [];
    for (let i = start; i < end; i++) {
        arr.push(i);
    }
    return arr;
}

function loadPgm(filename) {
    // Load an ASCII PGM (P2) into a 2D array.
    const lines = fs.readFileSync(filename, 'utf8')
        .split('\n')
        .filter(line => line.trim() && !line.startsWith('#'));

    if (lines[0].trim() !== 'P2') {
        throw new Error("Only ASCII PGM (P2) supported");
    }

    const [width, height] = lines[1].split(' ').map(Number);
    const pixels = lines.slice(3).join(' ').split(/\s+/).map(Number);

    const image = [];
    for (let i = 0; i < height; i++) {
        image.push(pixels.slice(i * width, (i + 1) * width));
    }

    return image;
}

/**
 * @Note Creates a list of [x-step, ..., x, ..., x+step]
 * @Assume @param step is a number, @param x can be a string
 */
function lrRange(x, step) {
  const res = Array.from({length: step*2 + 1}, (_,i) => {
    return i + Number(x)-step
  })
  return res
}

function avgPixel(img, y, x, sz) {
  const half = Math.floor(sz/2)

  let sumWindow = []

  let xlo = Math.max(x-half, 0)
  let xhi = Math.min(x+half+1, img[y].length)
  let row;          
  for (let irow of lrRange(y,half)) {
    if (irow < 0 || irow >= img.length) {
      row = []
    } 
    else {
      row = img[irow].slice(xlo, xhi)
    }
    sumWindow.push(row)
  if (sumWindow.length == 0) {
    return img[y][x]
  }
  let total = 0;

for (const row of sumWindow) {
    for (const value of row) {
        total += value;
    }
}
  return total / sumWindow.length**2
  }
}

function gaussianMultiplier(sz) {
  const half = Math.floor(sz/2)
  let weights = []
  let value = 1
  
  for (let i = 0; i < sz; i++) {
    weights.push(value)
    if (i < half) {
      value *= 2
    }  
    else {
      value = Math.floor(value/2)
    }
  }
  return weights
  
  }

function gaussianPixel(img, y, x, sz) {
  const half = Math.floor(sz/2)
  const weights = gaussianMultiplier(sz)
  const wd = img[0].length
  const ht = img.length
  let total = 0
  let weightSum = 0

  for (let dy of range(-half, half + 1)) {
    let imgy = y+dy
    if (imgy >= 0 && imgy < ht) {
      let wy = weights[dy + half]
      for (let dx of range (-half, half + 1)) {
        let imgx = x + dx
        if (ny >= 0 && ny < ht && nx >= 0 && nx < wd) {
          let wx = weights[dx + half]
          let w = wy * wx
          total += img[imgy][imgx] * w
          weightSum += w 
        }
      }
    }
  }
  let res = 0
  if (weightSum != 0) {
    return total / weightSum
  }
  else {
    return img[y][x]
  }
}

function pixeleval(img, template, y, x) {
  const sz = template.length
  const half = Math.floor(sz/2 )
  const ht = img.length
  const wd = img[0].length
  let res = 0
  let temp_sum = 0

  for (let yT of range(0, sz)) {
    for (let xT of range(0, sz)){

      let ny = y + yT - half
      let nx = x +xT - half

      if (ny >= 0 && ny < ht && nx >= 0 && nx < wd) {
        let w = template[yT][xT]
        res += img[ny][nx] * w
        temp_sum += w

      }
    }
  }
  if (temp_sum != 0) {
    return res / temp_sum
  } else { 
    return 0
  }
}

function imgProc(img, sz, operation = "avg", template=null) {
  let method = 0
  let res = []
  const ht = img.length
  const wd = img[0].length

  for (y in range(0, ht)) {
    res.push([])
    for (x in range(0, wd)) {
      if (template == null) {
        if (operation == "gaussian") {
         method = gaussianPixel(img, y, x, sz)
        }
        else if (operation == "avg") {
        method = avgPixel(img, y, x, sz)
  } 
}
        else {
         method = pixeleval(img, template, y, x)
}

        res[y].push(method);
      }
    }
    return res
  }
  


const readline = require('readline-sync')

function main() {

    const filename = readline.question("Enter PGM filename: ")
    const img = loadPgm(filename)

    while (true) {

        console.log("\n=== IMAGE PROCESSING MENU ===")
        console.log("1. Average Blur")
        console.log("2. Gaussian Blur")
        console.log("3. Custom Template")
        console.log("4. Quit")
        const choice = readline.question("Choose an option: ")

        if (choice === "1") {
            const sz = parseInt(readline.question("Kernel size: "))
            const result = imgProc(img, sz, "avg")
            console.log("\nProcessed Image:")
            console.log(result)
        }

        else if (choice === "2") {

            const sz = parseInt(readline.question("Kernel size: "))
            const result = imgProc(img, sz, "gaussian")
            console.log("\nProcessed Image:")
            console.log(result)
        } 

        else if (choice === "3") {
            const sz = parseInt(readline.question("Template size: "))
            console.log(`\nEnter ${sz * sz} numbers separated by spaces:`)
            let vals = readline.question("").split(" ").map(Number)
            // console.log('vals',vals)
            // vals = vals.split(" ").map(Number)
            const template = []

            for (let i = 0; i < sz; i++) {
                const row = vals.slice(i * sz, (i + 1) * sz)
                template.push(row)
            }

            console.log("\nTemplate:")
            for (const row of template) {
                console.log(row)
            }
            const result = imgProc(img, sz, "avg", template)
            console.log("\nProcessed Image:")
            console.log(result)
        }
        else if (choice === "4") {

            console.log("Goodbye")
            break
        }
        else {
            console.log("Invalid option")
        }
    }
}
main()